build/jemalloc.m4

```diff
@@ -19,59 +19,114 @@ dnl jemalloc.m4: Trafficserver's jemalloc autoconf macros
+
+  LDFLAGS="$save_ldflags"
+  CPPFLAGS="$save_cppflags"
+  LIBS="$(echo "$LIBS" | sed -e 's| *-ljemalloc||' -e 's|$| -ljemalloc|')"
+
+  AS_IF([test "x$have_jemalloc" == "xno" ], 
+     [AC_MSG_ERROR([Failed to compile with jemalloc.h and -ljemalloc]) ])
+
+  has_jemalloc=1
+dnl
+dnl flags must be global setting for all libs
+dnl 
+  TS_ADDTO(CPPFLAGS,$JEMALLOC_CFLAGS)
+  if test -n "$jemalloc_libdir"; then
+      TS_ADDTO_RPATH($jemalloc_libdir)
+      TS_ADDTO(LDFLAGS,$JEMALLOC_LDFLAGS)
   fi
-  if test "$jemalloc_have_headers" != "0"; then
-    jemalloch=1
-  else
-    CPPFLAGS=$saved_cppflags
-    LDFLAGS=$saved_ldflags
+
+  if test x$ac_cv_header_jemalloc_jemalloc_h = xyes; then
+      incpath="jemalloc/jemalloc.h";
+  elif test x$ac_cv_header_jemalloc_h = xyes; then
+      incpath="jemalloc.h";
   fi
-fi
-AC_SUBST(jemalloch)
+
+  cat > $ac_aux_dir/jemalloc_shim.cc <<END
+#include <$incpath>
+extern void* malloc(size_t n) noexcept __attribute__((weak));
+extern void free(void *p) noexcept __attribute__((weak));
+extern void* operator new(size_t n) __attribute__((weak));
+extern void* operator new@<:@@:>@(size_t n) __attribute__((weak));
+extern void operator delete(void* p) noexcept __attribute__((weak));
+extern void operator delete@<:@@:>@(void* p) noexcept __attribute__((weak));
+extern void operator delete(void* p, size_t n) noexcept __attribute__((weak));
+extern void operator delete@<:@@:>@(void* p, size_t n) noexcept __attribute__((weak));
+
+void* malloc(size_t n) throw() { return ( n ? mallocx(n,MALLOCX_ZERO) : NULL ); }
+void free(void *p) throw() { ( p ? dallocx(p,0) : (void)0 ); }
+void* operator new(size_t n) { return mallocx(n,MALLOCX_ZERO); }
+void* operator new@<:@@:>@(size_t n) { return mallocx(n,MALLOCX_ZERO); }
+void operator delete(void* p) noexcept { return dallocx(p,0); }
+void operator delete@<:@@:>@(void* p) noexcept { return dallocx(p,0); }
+void operator delete(void* p, size_t n) noexcept { return sdallocx(p,n,0); }
+void operator delete@<:@@:>@(void* p, size_t n) noexcept { return sdallocx(p,n,0); }
+END
+
+dnl  $SHELL $ac_abs_confdir/libtool --mode=compile $CXX -fPIC -DPIC -S -O2 $ac_aux_dir/jemalloc_shim.cc -c -o $ac_aux_dir/jemalloc_shim.s
+dnl
+dnl  wrapargs=" -Wl$($SED -e '${g;y/\n/,/;p}' -e '/^_Z.*:$/!d' -e 's/^\(.*\):/--wrap,\1/' -e 'H;d' $ac_aux_dir/jemalloc_shim.s)"
+dnl  TS_ADDTO(LDFLAGS,$wrapargs)
+dnl
+dnl  $SED -i -e 's/_Zn/__wrap_\0/g' -e 's/_Zd/__wrap_\0/g' $ac_aux_dir/jemalloc_shim.s
+dnl  $SHELL $ac_abs_confdir/libtool --mode=compile $CC -fPIC -DPIC -c -x assembler $ac_aux_dir/jemalloc_shim.s -o $ac_aux_dir/jemalloc_shim.lo
+
+  $SHELL $ac_abs_confdir/libtool --mode=compile $CXX -std=c++11 -fPIC -DPIC -O2 $ac_aux_dir/jemalloc_shim.cc -c -o $ac_aux_dir/jemalloc_shim.lo
+  $SHELL $ac_abs_confdir/libtool --mode=link $CC -o $ac_aux_dir/libjemalloc_shim.la $ac_aux_dir/jemalloc_shim.lo  
+
+  TS_ADDTO(LDFLAGS,$ac_abs_confdir/$ac_aux_dir/libjemalloc_shim.la)
```

Comment:
I think all the `$ac_aux_dir` here should be replaced with `$am_aux_dir`, otherwise it will fail if running `configure` in a sub-directory. The `$ac_abs_confdir/$ac_aux_dir/libjemalloc_shim.la` in line 126 should also be just `$am_aux_dir`.  Also, I have to add `$CPPFLAGS` to line 123 to make it work, otherwise it complains about cannot find the `jemalloc.h` file. Also, the `$ac_abs_confdir/libtool` should be replaced with just `` `pwd`/libtool``, and since the `libtool` script won't be generated till after you run the `configure`, so this will fail to generate the `libjemalloc_shim.la` file when running for the first time.

Reply:
I believe I merely chose that at the time.  Always the differences here.   

Both change 126 and 123?   For all your suggestions ... if they work on the Centos I'd prefer to change them to your way.    They were certainly hack-and-grind-until-working.

Reply:
I'm still trying to figure out a way to fix the libtool part, i'll push to the branch when I fix it.

Reply:
My sincerely apologies if master-jemalloc is messy or old.  I'd worked on it last about four months ago and I'd *thought* it was in good shape.   I'm working with the 6.2.x version now and it seems so sloppy.    [I'm working on a partial-object block storage plugin.. whcih is coming along well.]